Content.Server/Changeling/EntitySystems/ChangelingSystem.cs

```diff
@@ -0,0 +1,280 @@
+    private const string ChangelingRegenActionId = "ActionLingRegenerate";
+
+    [ValidatePrototypeId<EntityPrototype>]
+    private const string ChangelingAbsorbActionId = "ActionChangelingAbsorb";
+
+    [ValidatePrototypeId<EntityPrototype>]
+    private const string ChangelingDNACycleActionId = "ActionChangelingCycleDNA";
+
+    [ValidatePrototypeId<EntityPrototype>]
+    private const string ChangelingTransformActionId = "ActionChangelingTransform";
+
+    [ValidatePrototypeId<CurrencyPrototype>]
+    public const string EvolutionPointsCurrencyPrototype = "EvolutionPoints";
+
+    [ValidatePrototypeId<StorePresetPrototype>]
+    public const string ChangelingShopPresetPrototype = "StorePresetChangeling";
+
+    public bool ChangeChemicalsAmount(EntityUid uid, float amount, ChangelingComponent? component = null, bool regenCap = true)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+
+        component.Chemicals += amount;
+
+        if (regenCap)
+            float.Min(component.Chemicals, component.MaxChemicals);
+
+        _alerts.ShowAlert(uid, AlertType.Chemicals, (short) Math.Clamp(Math.Round(component.Chemicals / 10.7f), 0, 7));
+
+        return true;
+    }
+
+    private bool TryUseAbility(EntityUid uid, ChangelingComponent component, float abilityCost, bool activated = true, float regenCost = 0f)
+    {
+        if (component.Chemicals <= Math.Abs(abilityCost) && activated)
+        {
+            _popup.PopupEntity(Loc.GetString("changeling-not-enough-chemicals"), uid, uid);
+            return false;
+        }
+
+        if (activated)
+        {
+            ChangeChemicalsAmount(uid, abilityCost, component, false);
+            component.ChemicalsPerSecond -= regenCost;
+        }
+        else
+        {
+            component.ChemicalsPerSecond += regenCost;
+        }
+
+        return true;
+    }
+
+    private bool TryStingTarget(EntityUid uid, EntityUid target, ChangelingComponent component)
+    {
+        if (HasComp<ChangelingComponent>(target))
+        {
+            var selfMessage = Loc.GetString("changeling-sting-fail-self", ("target", Identity.Entity(target, EntityManager)));
+            _popup.PopupEntity(selfMessage, uid, uid);
+
+            var targetMessage = Loc.GetString("changeling-sting-fail-target");
+            _popup.PopupEntity(targetMessage, target, target);
+            return false;
+        }
+
+        return true;
+    }
+
+    private void OnMapInit(EntityUid uid, ChangelingComponent component, MapInitEvent args)
+    {
+        _action.AddAction(uid, ChangelingEvolutionMenuId);
+        _action.AddAction(uid, ChangelingRegenActionId);
+        _action.AddAction(uid, ChangelingAbsorbActionId);
+        _action.AddAction(uid, ChangelingDNACycleActionId);
+        _action.AddAction(uid, ChangelingTransformActionId);
+    }
+
+    private void OnShop(EntityUid uid, ChangelingComponent component, ChangelingEvolutionMenuActionEvent args)
```

Comment:
put this in its own comp and share it with revenant :trollface:

Reply:
i think this should be good now